File: ros2_self_msg_srv/src/my_interfaces_py/my_interfaces_py/publisher_person.py
#!/usr/bin/env python3
"""发布自定义消息示例 - Python版本"""
import rclpy
from rclpy.node import Node
from my_interfaces.msg import Person


class PublisherPerson(Node):
    """发布人员信息节点"""

    def __init__(self):
        super().__init__('publisher_person')
        # 创建发布者，发布自定义消息Person
        self.publisher_ = self.create_publisher(Person, 'person_info', 10)
        
        # 创建定时器，每1秒发布一次
        self.timer = self.create_timer(1.0, self.timer_callback)
        self.count = 0
        
        self.get_logger().info('发布者已启动，发布人员信息...')

    def timer_callback(self):
        """定时器回调：创建并发布消息"""
        msg = Person()
        msg.name = f'张三_{self.count}'
        msg.age = 25 + (self.count % 10)
        msg.height = 1.75 + (self.count % 5) * 0.01
        
        self.publisher_.publish(msg)
        self.get_logger().info(
            f'发布: 姓名={msg.name}, 年龄={msg.age}, 身高={msg.height:.2f}m'
        )
        self.count += 1


def main(args=None):
    rclpy.init(args=args)
    publisher = PublisherPerson()
    try:
        rclpy.spin(publisher)
    except KeyboardInterrupt:
        pass
    finally:
        publisher.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_self_msg_srv/src/my_interfaces_cpp/src/publisher_person.cpp
/**
 * 发布自定义消息示例 - C++版本
 */
#include "rclcpp/rclcpp.hpp"
#include "my_interfaces/msg/person.hpp"

using std::placeholders::_1;

class PublisherPerson : public rclcpp::Node
{
public:
    PublisherPerson() : Node("publisher_person"), count_(0)
    {
        // 创建发布者，发布自定义消息Person
        publisher_ = this->create_publisher<my_interfaces::msg::Person>("person_info", 10);
        
        // 创建定时器，每1秒发布一次
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&PublisherPerson::timer_callback, this));
        
        RCLCPP_INFO(this->get_logger(), "发布者已启动，发布人员信息...");
    }

private:
    void timer_callback()
    {
        auto msg = my_interfaces::msg::Person();
        msg.name = "李四_" + std::to_string(count_);
        msg.age = 30 + (count_ % 10);
        msg.height = 1.80 + (count_ % 5) * 0.01;
        
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), 
            "发布: 姓名=%s, 年龄=%d, 身高=%.2fm",
            msg.name.c_str(), msg.age, msg.height);
        count_++;
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<my_interfaces::msg::Person>::SharedPtr publisher_;
    size_t count_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<PublisherPerson>());
    rclcpp::shutdown();
    return 0;
}
